### ISSUE TITLE
feat(cvm): finalize lowering routing algorithm spec

### ISSUE THREAD

**Issue by dzmitry-lahoda** [ComposableFi/composable#4395]
# Algorithm problem about transactions routing.

https://github.com/ComposableFi/composable/blob/dz/c/6/docs/docs/technology/cvm/lowering-routing.md

Algorithm is to be plugged somewhere here https://github.com/ComposableFi/cvm/blob/6e7500827fd65d8993c781a03e99dcfbb2e5213b/contracts/cosmwasm/outpost/src/contract/ibc/ics20.rs#L32

## All chatter around 

hello there,  I am thinking how to route cross chain "packets" over hetergenous networks with various capabilities. So I have tree AST(CVM program describing what to do). On chain Registry(Graph describing what features and connections exists, may be partial info), and real whole networks with features from doing forwarding only to full blown contracts (yeah - there are chains with no contracts and supporting one and only one token).  I need some algorithm to down from AST to some level route with description what transport to use. Here is some description with examples.

dzmitry lahoda, [28/12/2023 10.01]
Hello

dzmitry lahoda, [28/12/2023 10.01]
https://github.com/ComposableFi/composable/pull/4385/files#diff-ab77bafbc5733029d7fff672eaf10a837d0f8a04b51426fa705c54ef18109060

dzmitry lahoda, [28/12/2023 10.02]
TLDR; CVM program is high level, best when there are contracts on each and chains are cheap

dzmitry lahoda, [28/12/2023 10.04]
but when underlying bridge(connectivity) is heterehenous(no contracts, different bridge features) with price of some chains high(ethereum), need to lower CVM progam directly into underlying execution "langauge" depending on features. CVM fully on chain now, assuming cheap chain to do some tree search, but I do not have estimate now

dzmitry lahoda, [28/12/2023 10.04]
CVM described in this file https://github.com/ComposableFi/composable/pull/4385/files#diff-91ad604ca0dac56d69a9e7110a987427e98e17cc1b4712330ae4414ef3a960f9 , it is semi formal so

dzmitry lahoda, [28/12/2023 10.05]

## Possible ideas of implementation

CVM is tree.

Registry is graph.

dzmitry lahoda, [28/12/2023 10.05]
My next step could be to write algorithm in python, and answer ## Solution constraints and questions

things with current draft document about CVM routing(and abi) lowering is:
0. sure need to verify/change/improve existing description of pseudocode (either via python or via gas). because it will take 20x more code to code it in rust.
some chains cannot do it on chain for sure. what is lowering price in gas?
can any chain do it eventually at all? (any hop at all)
what is impact on partial knowledge?
how to get prices on chain? (need may happen sooner than later)
given this, would not be CVM router be off chain? (actually on chain only adapters/keeprs/outsposts for execution/tracking + registry; but CVM as language is purely offchain).
i feel that CVM to be offchain. but i do not have proves.

12 replies
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703842932886649?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
all these mentioned in doc
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703843017060359?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
on chain is bad for now for simple reason - redeploy of contacts is too often, audit is hard, need to rewrite later
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703843075239509?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
but with some nice hack - run router part CW VM in browser or on chain
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703843084528309?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
so that same code operates both
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703843097555869?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
that reuturns the need to return router contract in different form
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703843101713829?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
but that is detail
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703843136219629?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
but should suffer CW or not for the sake of now on chain somewhere or not
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703843603683999?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
if offchain
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703843607842429?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
centralized or not
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703843635604419?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
if centralized - should be done in python, cause I have all data in web server and already output of algorithm
Dzmitry Lahoda
  [5 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703844249398679?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
this decision should have been taken 1 year ago
Dzmitry Lahoda
  [4 hours ago](https://composableprotocol.slack.com/archives/C05EMF8DYKU/p1703849537725949?thread_ts=1703842888.975909&cid=C05EMF8DYKU)
so offchain cannot verify that CVM intention was properly converted to route. sure single user execution can be done by some TS code in browser. but for batch autions, it will increase trust to solvers (they already fully trusted).

**Comment by dzmitry-lahoda** [comment]
problem described several ways, nees tp review it ans write algorithms
